src/Storages/Statistics/Statistics.cpp

```diff
@@ -71,17 +71,29 @@ void ColumnStatistics::update(const ColumnPtr & column)
         stat.second->update(column);
 }
 
-UInt64 IStatistics::estimateCardinality() const
+void ColumnStatistics::merge(const ColumnStatistics & other)
+{
+    rows += other.rows;
+    for (const auto & [type, stat] : other.stats)
+    {
+        if (!stats.contains(type))
+            stats[type] = stat;
+        else
+            stats[type]->merge(stat);
+    }
```

Comment:
So l. 76ff are doing what `ConditionSelectivityEstimator::merge` and `ConditionSelectivityEstimator::ColumnSelectivityEstimator::merge` were doing, right?

Two questions:
- I am confused about l. 82. It calls the new `merge` function. In the previous code, `ConditionSelectivityEstimator::ColumnSelectivityEstimator::merge` threw a LOGICAL_ERROR if the statistics were already there. How can we even run into this situation?
- I like to understand what statistics are we actually merging? Let's say the user defined minmax and tdigest statistics for a column and there are three parts. So we would basically merge the three min-max statistics together, right? Does that deteriorate the quality of the estimation somehow? Let's say the [min, max] statistics are [1, 3], [77, 80], [182, 189] in a column that contains integers between 0 and 200. Individually, they provide very expressive and meaningful statistics about the content of each part (each part contains a very narrow range of values). If we merge them together we get [1, 189] which is basically useless as a statistics, I guess?

Reply:
> So l. 76ff are doing what ConditionSelectivityEstimator::merge and ConditionSelectivityEstimator::ColumnSelectivityEstimator::merge were doing, right?

yes, but in different fashion, see https://github.com/ClickHouse/ClickHouse/pull/69380#discussion_r1758151574

>  Let's say the user defined minmax and tdigest statistics for a column and there are three parts. So we would basically merge the three min-max statistics together, right?

yes

> Does that deteriorate the quality of the estimation somehow?

yes, this indeed will deteriorate the quality of the estimation somehow. The main reason I introduce the merging is 
1. memory usage: for some statistics cost large memory When a query hits a large number of parts, it will consume a lot of memory.
2. Statistics have a wide range of applications in databases, such as CBO, actually I am working on a CBO optimizer which we ever submited as a PR #56827, that is why I submitted the statistics related prs. Part-level statistics are too granular and not well-suited for many scenarios.

haha We need a trade-off to make a choice.

Reply:
> Does that deteriorate the quality of the estimation somehow?

Kinda unrelated, but there is "better" variant of minmax statistics for exactly this kind of situations https://github.com/ClickHouse/ClickHouse/issues/56341

Reply:
> > Does that deteriorate the quality of the estimation somehow?
> 
> Kinda unrelated, but there is just "better" variant of minmax statistics for exactly this kind of situations #56341

BTW for statistics they store not only bins but also extremes. Like a column contains strings and they somehow distributed but also some string appeared with maximum frequency, like:

bin1: [Armadillo ... Camel] = 3005676 (size of the bin)
bin2: [Chimaera... Gorilla] = 3001000 (size of the bin)
bin3: [Grizzly Bear... Rat] = 3101123 (size of the bin)
...
extreme1: _= 5554786778
extreme2: null= 155478677856
extreme3: cat= 114786778
extreme4: dog= 212755578

And number of bins and extremes are limited by some reasonable limit, like 20 - bins, 5 - extremes

Reply:
`histogram` is an advanced statistics than `minmax` for range condition, it can mitigate the deterioration caused by merging. @UnamedRus @den-crane thanks for the point.

`histogram` is also worthy to introduce as a statistics type (@den-crane proposed it as a skip index), although there is a similar one `TDigest`. 

1. `histogram` provides more accuracy estimation, because `histogram` provides boundries and we can calculate the proportion of the range of the condition within the hit bins. 
2. `histogram` is calculation friendly. For example: `col > 100 and col < 200`: after applying the expression `col > 100`, the histogram is trimmed to exclude data less than 100, and then used as input for calculating `col < 200` and then get the final histogram.

Reply:
We couldn't say which data structure is better before we actually use it to estimate something.

Can we use the current statistics to estimate something? It's not hard to add/remove/modify some types of statistics.

Reply:
Actually I didn't understand the idea of merging statistics. Why do we need it for CBO? AFAIU Estimator is only a block box, CBO should not care about it

Reply:
Regarding histograms: The good thing of the existing statistics code is that it is extensible. We can add histograms in any shape or form as a new statistics type later. But please do such discussions in a separate PR, not here.

Regarding the merging of statistics during estimation:

> memory usage: for some statistics cost large memory When a query hits a large number of parts, it will consume a lot of memory.

I would say the point of _any_ statistics type is to be light-weight in terms of memory. Yes, a large number of parts could lead to a high total memory consumption of the corresponding statistics but that is also relative. Assume a (pessimistic) memory size of 10 kB per statistics object. With 1000 parts (also a pessimistic assumption), we get 10 MB statistics size in total. That is totally okay given that servers have dozens of GB of RAM, typically.

> Statistics have a wide range of applications in databases, such as CBO, actually I am working on a CBO optimizer which we ever submited as a PR 
> https://github.com/ClickHouse/ClickHouse/pull/56827, that is why I submitted the statistics related prs. Part-level statistics are too granular and not well-suited for many scenarios.

One design principle of ClickHouse is to have _no_ "global" data structures. By "global", I mean everything that goes beyond part granularity. For example, primary key indexes are per-part, projections are per part, skipping indexes are per part, inserts create new parts (but never add to, i.e. modify, existing parts), etc. Avoiding global data structures avoids lock contention, this helps with scalability of the entire database.

I would say we should maintain this principle for statistics already for conceptual reasons. And note that this does not make the math more difficult. Think about this: Statistics allow to estimate either the cardinality of a column in a part (`estimateCardinality`), or the selectivity of a column in a part under some predicate (`estimateEqual`, `estimateLess`). In both cases, a row count is returned. So with N parts, we'll obtain N such row counts. For every part, we know the total number of rows. Computing the column-level row count (across all parts) is then as easy as normalizing the estimated row counts per part by the part's respective total row count.

Besides that, I still think that merging statistics potentially deteriorates the quality of estimations as demonstrated by the min-max example above (which may be an extreme case but such situations will definitely occur in practice).

Reply:
> Actually I didn't understand the idea of merging statistics. Why do we need it for CBO? AFAIU Estimator is only a block box, CBO should not care about it

Indeed CBO and other scenarios should not care about estimation. The reason I mention it is that the statistics will not to be too granular.

Reply:
@rschu1ze thanks for your explaination, I will revert the merging logic. However, there's one point I'd like to raise.

Every node in an `ActionsDag` will calculate with the statistics, for example `a > 100 and a < 200 and (c in (1, 2, 3) or c > 20) and d = 'xx' ...`, there will be a calculation for `a`, `a > 100`, `a > 100 and a<200` ..., there can be a lot of statistics calculation.  If we have 1000 parts every calculation will handle 1000 statistics.

Reply:
> Can we use the current statistics to estimate something? 

The current statistics is enough for me.

Reply:
@JackyWoo Yes there is always risk to calculate complex condition, we can optimize it when we actually meet this problem, I think that's should not be hard

Reply:
@JackyWoo Regarding your points:

> 1. Every node in an ActionsDag will calculate with the statistics, for example a > 100 and a < 200 and (c in (1, 2, 3) or c > 20) and d = 'xx' ..., there will be a calculation for a, a > 100, a > 100 and a<200 .... If we have 1000 parts every calculation will handle 1000 statistics. so there can be a lot of statistics calculations.

I'd consider this behavior a bug in the logic that calls the query estimation. Given a complex condition like `a > 100 and a < 200 and (c in (1, 2, 3) or c > 20) and d = 'xx'` a sane optimizer would compute the disjunctive or conjunctive normal form of the full conditions, and then estimate each condition individually (but not columns like `a`) to decide in which order to evaluate them.

Maybe doing such estimations in ActionsDAG is already too deep in the stack ...

Reply:
As my understanding the logic you described is to order the condition. What about if want to estimate the whole condition(a cob need )? We should traverse the ActionsDag.

please correct me if if my understanding is wrong.